### PR TITLE
recursive synchronous handler blocks on ThreadPoolExecutor::submit().

### PR DESCRIPTION
**Pull request by fujitatomoya**
address https://github.com/ros2/rclpy/issues/1016

Signed-off-by: Tomoya Fujita <Tomoya.Fujita@sony.com>

### CODE REVIEW

**Comment by fujitatomoya** [comment]
This is draft PR to get feedback since this regenerates the problem https://github.com/ros2/rclpy/issues/983.

**Comment by fujitatomoya** [comment]
@wjwwood @ivanpauno @sloretz any thoughts? asking for help...

**Comment by wjwwood** [comment]
Is it possible to make a regression test for the issue in https://github.com/ros2/rclpy/issues/1016? That would be the best way to prove that fixes the problem.

**Comment by fujitatomoya** [comment]
> Is it possible to make a regression test for the issue in https://github.com/ros2/rclpy/issues/1016?

yeah, i think so. Using `MultiThreadedExecutor` to call sync client request via timer callback will reproduce the issue.

But before that, i do not have good idea to address https://github.com/ros2/rclpy/issues/1016 and https://github.com/ros2/rclpy/issues/983 atm...

**Comment by fujitatomoya** [comment]
CC: @Barry-Xu-2018 @iuhilnehc-ynos

**Review by ivanpauno — APPROVED** [review]
Makes sense!

**Comment by ivanpauno** [comment]
> But before that, i do not have good idea to address https://github.com/ros2/rclpy/issues/1016 and https://github.com/ros2/rclpy/issues/983 atm

aren't both issues fixed after this PR?

**Comment by fujitatomoya** [comment]
> aren't both issues fixed after this PR?

No, unfortunately. This PR regenerates https://github.com/ros2/rclpy/issues/983, that is why I keep this under draft.

File: rclpy/rclpy/executors.py [hunk 1]
```diff
@@ -769,7 +770,7 @@ def handler_wrapper(handler):
                 handler()
                 if handler.exception() is not None:
                     raise handler.exception()
-            self._executor.submit(handler_wrapper(handler))
+            self._executor.submit(partial(handler_wrapper, handler))
```
Comment (fujitatomoya) [inline rclpy/rclpy/executors.py:773]: This also works the same.

```suggestion
            self._executor.submit(handler_wrapper, handler)
```

Comment (ivanpauno) [reply rclpy/rclpy/executors.py:773]: Well, I didn't notice it in my original https://github.com/ros2/rclpy/pull/984 review, but that PR is broken.
It's not submiting `handler` for execution in the thread pool.
It's execution the handler immediately and submitting `None` to the executor ...

The only way to detect exceptions is to store the future returned by the executor and check if they finished successfully or with an exception.
We cannot block until the future is completed, so in each iteration of `spin()` we could poll all the stored futures and check if one of them failed (removing the ones that completed).
When a failure is detecter, it's thrown.

---

I'm okay with reverting https://github.com/ros2/rclpy/pull/984 temporally, as the solution to the actual problem #984 tried to solve isn't that trivial.

Comment (ivanpauno) [reply rclpy/rclpy/executors.py:773]: @wjwwood what do you think?

Comment (fujitatomoya) [reply rclpy/rclpy/executors.py:773]: > I'm okay with reverting https://github.com/ros2/rclpy/pull/984 temporally, as the solution to the actual problem https://github.com/ros2/rclpy/pull/984 tried to solve isn't that trivial.

agree.

Comment (ivanpauno) [reply rclpy/rclpy/executors.py:773]: BTW, thanks for the PR @fujitatomoya !
I didn't undestand what was the issue until this :sweat_smile:

Comment (fujitatomoya) [reply rclpy/rclpy/executors.py:773]: Reopen revert commit https://github.com/ros2/rclpy/pull/1017

**Comment by fujitatomoya** [comment]
I will close this draft, instead will do the following.

- https://github.com/ros2/rclpy/pull/1017 (revert https://github.com/ros2/rclpy/pull/984) to address https://github.com/ros2/rclpy/issues/1016
- Backport https://github.com/ros2/rclpy/pull/1017 to humble / galactic / foxy
- Reopen https://github.com/ros2/rclpy/issues/983, will address with new PR. (Suggested fix, see https://github.com/ros2/rclpy/pull/1019#discussion_r988178727)
- Give it some time window before backport humble / galactic / foxy at this time.

**Pull request closed without merging**